File: catkin_ws_vps_p/src/kf_mpc_controller4/src/state_predictor.cpp
#include "../include/kf_mpc_controller4/state_predictor.h"

state_predictor::state_predictor()
{
  ros::NodeHandle n("~");
  n.getParam("predict_positiong_pub_topic",_predict_positiong_pub_topic);
  n.getParam("state_predictor_info_pub_topic",_state_predictor_info_pub_topic);
  n.getParam("system_state_sub_topic",_system_state_sub_topic);
  n.getParam("predictor_run_hz",_predictor_run_hz);

  n.getParam("predict_time_dt_sub_topic",_predict_time_dt_sub_topic);

  n.getParam("save_data_window",_save_data_window);
  n.getParam("max_predict_data_window",_max_predict_data_window);
  n.getParam("min_predict_data_window",_min_predict_data_window);
  n.getParam("predict_window",_predict_window);
  n.getParam("predict_order",_predict_order);
  n.getParam("predict_time_dt",_predict_time_dt);
  n.getParam("queue_check_error",_queue_check_error);

  n.getParam("max_velocity_x",_max_velocity_x);
  n.getParam("max_velocity_y",_max_velocity_y);
  n.getParam("max_velocity_z",_max_velocity_z);

  _system_state_sub = n.subscribe(_system_state_sub_topic,1,&state_predictor::system_state_sub_cb,this);
  _predict_time_dt_sub = n.subscribe(_predict_time_dt_sub_topic,1,&state_predictor::predict_time_dt_sub_cb,this);
  _predict_position_pub = n.advertise<dt_message_package::object_predict_position>(_predict_positiong_pub_topic,1);
  _state_predictor_info_pub = n.advertise<dt_message_package::state_predictor_info>(_state_predictor_info_pub_topic,1);

  _cicl_queue_t.resize(_save_data_window);
  _cicl_queue_x.resize(_save_data_window);
  _cicl_queue_y.resize(_save_data_window);
  _cicl_queue_z.resize(_save_data_window);


  _predictor_x = new least_square();
  _predictor_y = new least_square();
  _predictor_z = new least_square();

  _predictor_x->set_Order(_predict_order);
  _predictor_y->set_Order(_predict_order);
  _predictor_z->set_Order(_predict_order);

  _predict_data_x_window = _max_predict_data_window;
  _predict_data_y_window = _max_predict_data_window;
  _predict_data_z_window = _max_predict_data_window;

  _predict_time_dt_x = _predict_time_dt;
  _predict_time_dt_y = _predict_time_dt;
  _predict_time_dt_z = _predict_time_dt;

  _x_event_id = 1;
  _y_event_id = 1;
  _z_event_id = 1;

  _running_data_x_t.resize(_predict_data_x_window);
  _running_data_y_t.resize(_predict_data_y_window);
  _running_data_z_t.resize(_predict_data_z_window);
  _running_data_x.resize(_predict_data_x_window);
  _running_data_y.resize(_predict_data_y_window);
  _running_data_z.resize(_predict_data_z_window);

  int flag_thread = pthread_create(&_predictor_run_thread,NULL,&state_predictor::predictor_running,this);
  if (flag_thread < 0)
  {
    ROS_ERROR("pthread_create ros_process_thread failed: %d\n", flag_thread);
  }

}

void state_predictor::system_state_sub_cb(const dt_message_package::object_move_stateConstPtr &msg)
{
  std::lock_guard<std::mutex> lock(_data_mutex);
  _cicl_queue_t.push_back(msg.get()->curren_time);
  _cicl_queue_x.push_back(msg.get()->position_x);
  _cicl_queue_y.push_back(msg.get()->position_y);
  _cicl_queue_z.push_back(msg.get()->position_z);
}

void state_predictor::predict_time_dt_sub_cb(const dt_message_package::predict_time_dtConstPtr &msg)
{
  _predict_time_dt_x = msg.get()->predict_time_dt_x;
  _predict_time_dt_y = msg.get()->predict_time_dt_y;
  _predict_time_dt_z = msg.get()->predict_time_dt_z;

}
predict_state_stu state_predictor::static_motion(float current_times,std::vector<float> state,int predict_window,float predict_time_dt)
{
  predict_state_stu ret;
  ret.states.resize(predict_window);
  ret.times.resize(predict_window);
  float current_state = state.at(state.size()-1);
  for(int i =0;i<predict_window;i++)
  {
    ret.times.at(i) = (i+1)*predict_time_dt + current_times;
    ret.states.at(i) = current_state;
  }
  return ret;
}

predict_state_stu state_predictor::max_velocity_motion(float current_times,std::vector<float> state, int predict_window,float predict_time_dt,float max_velocity)
{
  float move_dir = state.at(state.size()-1) - state.at(state.size()-2);
  float move_velocity = 0;
  if(move_dir > 0.005)
  {
    move_velocity = max_velocity;
  }
  else if(move_dir < 0.005)
  {
    move_velocity = -max_velocity;
  }
  else
  {
    move_velocity = 0;
  }
  //std::cout<<"move_velocity: "<<move_velocity<<std::endl;
  predict_state_stu ret;
  ret.times.resize(predict_window);
  ret.states.resize(predict_window);
  float current_state = state.at(state.size()-1);
  for(int i = 0; i<predict_window; i++)
  {
    ret.times.at(i) = current_times + (i+1)*predict_time_dt;
    ret.states.at(i) = current_state + (i+1)*move_velocity * predict_time_dt;
  }
  return ret;
}
predict_state_stu state_predictor::least_squares_motion(least_square* least_square_predictor,float current_times, std::vector<float> times,std::vector<float> state,int predict_window,float predict_time_dt)
{
  least_square_predictor->set_T(times);
  least_square_predictor->set_X(state);
  Eigen::VectorXf result = least_square_predictor->get_Result2();
  predict_state_stu ret;
  ret.times.resize(predict_window);
  ret.states.resize(predict_window);
  for(int i = 0; i< predict_window; i++)
  {
    float predict_time = current_times + (i+1)*predict_time_dt;
    ret.times.at(i) = predict_time;
    for(int j = result.size()-1;j>=0;j--)
    {
      float predict_time = times.at(times.size()-1) + (i+1)*predict_time_dt;
      ret.states.at(i) = ret.states.at(i) + pow(predict_time,result.size()-1-j)*result[j];
    }
  }
  return ret;
}

double state_predictor::GetMinValue(std::vector<float> data_queue)
{
 // for(int i = 0;i<data_queue.size();i++)
 // {
 //     std::cout<<"data_queue: "<<data_queue.at(i)<<std::endl;
 // }

  float min_value = 100.0;
  for(int i =0;i<data_queue.size();i++)
  {
    if(data_queue.at(i)<min_value)
    {
      min_value = data_queue.at(i);
    }
  }
  return min_value;
}
double state_predictor::GetMaxValue(std::vector<float> data_queue)
{
  float max_value = 0.0;
  for(int i =0;i<data_queue.size();i++)
  {
    if(data_queue.at(i)>max_value)
    {
      max_value = data_queue.at(i);
    }
  }
  return max_value;
}

int state_predictor::event_trigger(std::vector<float> data_queue,float check_err)
{
  std::vector<float> dev_data_queue;
  int event_id = 0;
  dev_data_queue.resize(data_queue.size()-1);
  for(int i = 1;i<data_queue.size();i++)
  {
    dev_data_queue.at(i-1) = fabs(data_queue.at(i) - data_queue.at(i-1));
  }

  double min_value = GetMinValue(dev_data_queue);
  double max_value = GetMaxValue(dev_data_queue);
  if (max_value>check_err && min_value>check_err)
  {
    event_id = 0;
  }
  else if(max_value<=check_err && min_value<= check_err)
  {
    event_id = 1;
  }
  else
  {
    event_id = 2;
  }
  //std::cout<<"min_value: "<<min_value<<std::endl;
  //std::cout<<"max_value: "<<max_value<<std::endl;
  return event_id;

}

int state_predictor::get_predict_window_size(int current_event_id, int last_predict_window_size, int max_window_size, int min_window_size)
{
  int now_predict_window_size = last_predict_window_size;
  if (current_event_id == 0)
  {
    now_predict_window_size = now_predict_window_size + 1;
  }
  else if (current_event_id == 1)
  {
    now_predict_window_size = now_predict_window_size;
  }
  else
  {
    now_predict_window_size = min_window_size;
  }

  if (now_predict_window_size > max_window_size)
  {
    now_predict_window_size = max_window_size;
  }
  else if (now_predict_window_size < min_window_size)
  {
    now_predict_window_size = min_window_size;
  }
  else
  {
    now_predict_window_size = now_predict_window_size;
  }

  return now_predict_window_size;
}

void *state_predictor::predictor_running(void *args)
{
  state_predictor* predictor = (state_predictor*)(args);

  ros::Rate rate(predictor->_predictor_run_hz);
  dt_message_package::object_predict_position object_pre_position;
  object_pre_position.times.resize(predictor->_predict_window);
  object_pre_position.position_xs.resize(predictor->_predict_window);
  object_pre_position.position_ys.resize(predictor->_predict_window);
  object_pre_position.position_zs.resize(predictor->_predict_window);
  while(ros::ok())
  {
    if(predictor->_cicl_queue_t.at(predictor->_cicl_queue_t.size()-1)!=0)
    {
      int j = 0;
      double current_time = predictor->_cicl_queue_t.at(predictor->_cicl_queue_t.size()-1);
      float current_position_x = predictor->_cicl_queue_x.at(predictor->_cicl_queue_x.size()-1);
      float current_position_y = predictor->_cicl_queue_y.at(predictor->_cicl_queue_y.size()-1);
      float current_position_z = predictor->_cicl_queue_z.at(predictor->_cicl_queue_z.size()-1);

      predictor->_running_data_x_t.resize(predictor->_predict_data_x_window);
      predictor->_running_data_y_t.resize(predictor->_predict_data_y_window);
      predictor->_running_data_z_t.resize(predictor->_predict_data_z_window);
      predictor->_running_data_x.resize(predictor->_predict_data_x_window);
      predictor->_running_data_y.resize(predictor->_predict_data_y_window);
      predictor->_running_data_z.resize(predictor->_predict_data_z_window);

      for(int i = predictor->_cicl_queue_t.size()-predictor->_predict_data_x_window;i<predictor->_cicl_queue_t.size();i++)
      {
        predictor->_running_data_x_t.at(j) = predictor->_cicl_queue_t.at(i) - predictor->_cicl_queue_t.at(predictor->_cicl_queue_t.size()-predictor->_predict_data_x_window);
        predictor->_running_data_x.at(j) = predictor->_cicl_queue_x.at(i);
       // std::cout<<"running_data_x_t: "<<j<<": "<<predictor->_running_data_x_t.at(j)<<std::endl;
       // std::cout<<"running_data_x: "<<j<<": "<<predictor->_running_data_x.at(j)<<std::endl;
        // std::cout<<"_cicl_queue_t: "<<j<<": "<<predictor->_cicl_queue_t.at(i)<<std::endl;

        j = j + 1;
      }
      j = 0;

      for(int i = predictor->_cicl_queue_t.size()-predictor->_predict_data_y_window;i<predictor->_cicl_queue_t.size();i++)
      {
        predictor->_running_data_y_t.at(j) = predictor->_cicl_queue_t.at(i) - predictor->_cicl_queue_t.at(predictor->_cicl_queue_t.size()-predictor->_predict_data_y_window);
        // std::cout<<"running_data_t: "<<j<<": "<<predictor->_running_data_t.at(j)<<std::endl;
        //  std::cout<<"_cicl_queue_t: "<<j<<": "<<predictor->_cicl_queue_t.at(i)<<std::endl;
        predictor->_running_data_y.at(j) = predictor->_cicl_queue_y.at(i);
        j = j + 1;
      }
      j = 0;

      for(int i = predictor->_cicl_queue_t.size()-predictor->_predict_data_z_window;i<predictor->_cicl_queue_t.size();i++)
      {
        predictor->_running_data_z_t.at(j) = predictor->_cicl_queue_t.at(i) - predictor->_cicl_queue_t.at(predictor->_cicl_queue_t.size()-predictor->_predict_data_z_window);
        // std::cout<<"running_data_t: "<<j<<": "<<predictor->_running_data_t.at(j)<<std::endl;
        //  std::cout<<"_cicl_queue_t: "<<j<<": "<<predictor->_cicl_queue_t.at(i)<<std::endl;
        predictor->_running_data_z.at(j) = predictor->_cicl_queue_z.at(i);
        j = j + 1;
      }

      predictor->_x_event_id = predictor->event_trigger( predictor->_running_data_x,predictor->_queue_check_error);
      predictor->_y_event_id = predictor->event_trigger( predictor->_running_data_y,predictor->_queue_check_error);
      predictor->_z_event_id = predictor->event_trigger( predictor->_running_data_z,predictor->_queue_check_error);

    //  std::cout<<"x_event_id: "<<predictor->_x_event_id<<std::endl;
    //  std::cout<<"y_event_id: "<<predictor->_y_event_id<<std::endl;
    //  std::cout<<"z_event_id: "<<predictor->_z_event_id<<std::endl;

      predictor->_predict_data_x_window =  predictor->get_predict_window_size( predictor->_x_event_id, predictor->_predict_data_x_window, predictor->_max_predict_data_window, predictor->_min_predict_data_window);
      predictor->_predict_data_y_window =  predictor->get_predict_window_size( predictor->_y_event_id, predictor->_predict_data_y_window, predictor->_max_predict_data_window, predictor->_min_predict_data_window);
      predictor->_predict_data_z_window =  predictor->get_predict_window_size( predictor->_z_event_id, predictor->_predict_data_z_window, predictor->_max_predict_data_window, predictor->_min_predict_data_window);

      predict_state_stu pdt_state_x_stu;
      predict_state_stu pdt_state_y_stu;
      predict_state_stu pdt_state_z_stu;

      switch (predictor->_x_event_id) {
      case 0:
      {
        pdt_state_x_stu = predictor->least_squares_motion(predictor->_predictor_x,current_time, predictor->_running_data_x_t,predictor->_running_data_x,predictor->_predict_window,predictor->_predict_time_dt_x);
      }
        break;
      case 1:
      {
        pdt_state_x_stu = predictor->static_motion(current_time,predictor->_running_data_x,predictor->_predict_window,predictor->_predict_time_dt_x);
      }
        break;
      case 2:
      {
        pdt_state_x_stu = predictor->max_velocity_motion(current_time,predictor->_running_data_x, predictor->_predict_window,predictor->_predict_time_dt_x,predictor->_max_velocity_x);
      }
        break;
      default:
        break;
      }

      switch (predictor->_y_event_id) {
      case 0:
      {
        pdt_state_y_stu = predictor->least_squares_motion(predictor->_predictor_y,current_time, predictor->_running_data_y_t,predictor->_running_data_y,predictor->_predict_window,predictor->_predict_time_dt_y);
      }
        break;
      case 1:
      {
        pdt_state_y_stu = predictor->static_motion(current_time,predictor->_running_data_y,predictor->_predict_window,predictor->_predict_time_dt_y);
      }
        break;
      case 2:
      {
        pdt_state_y_stu = predictor->max_velocity_motion(current_time,predictor->_running_data_y, predictor->_predict_window,predictor->_predict_time_dt_y,predictor->_max_velocity_y);
      }
        break;
      default:
        break;
      }

      switch (predictor->_z_event_id) {
      case 0:
      {
        pdt_state_z_stu = predictor->least_squares_motion(predictor->_predictor_z,current_time, predictor->_running_data_z_t,predictor->_running_data_z,predictor->_predict_window,predictor->_predict_time_dt_z);
      }
        break;
      case 1:
      {
        pdt_state_z_stu = predictor->static_motion(current_time,predictor->_running_data_z,predictor->_predict_window,predictor->_predict_time_dt_z);
      }
        break;
      case 2:
      {
        pdt_state_z_stu = predictor->max_velocity_motion(current_time,predictor->_running_data_z, predictor->_predict_window,predictor->_predict_time_dt_z,predictor->_max_velocity_z);
      }
        break;
      default:
        break;
      }

      for(int i = 0; i< pdt_state_x_stu.states.size();i++)
      {
        object_pre_position.times.at(i) = pdt_state_x_stu.times.at(i);
        object_pre_position.position_xs.at(i) = pdt_state_x_stu.states.at(i);
        object_pre_position.position_ys.at(i) = pdt_state_y_stu.states.at(i);
        object_pre_position.position_zs.at(i) = pdt_state_z_stu.states.at(i);
      }
      object_pre_position.current_position_x = current_position_x;
      object_pre_position.current_time = current_time;
      object_pre_position.current_position_y = current_position_y;
      object_pre_position.current_position_z = current_position_z;
      predictor->_predict_position_pub.publish(object_pre_position);
      //=========================================================out state predictor information=========================================
      {
       dt_message_package::state_predictor_info predictor_info;
       predictor_info.system_time = ros::Time::now().toNSec();
       predictor_info.predict_window = predictor->_predict_window;
       predictor_info.max_predict_data_window = predictor->_max_predict_data_window;
       predictor_info.min_predict_data_window = predictor->_min_predict_data_window;
       predictor_info.predict_time_dt_x = predictor->_predict_time_dt_x;
       predictor_info.predict_time_dt_y = predictor->_predict_time_dt_y;
       predictor_info.predict_time_dt_z = predictor->_predict_time_dt_z;
       predictor_info.predict_data_x_window = predictor->_predict_data_x_window;
       predictor_info.predict_data_y_window = predictor->_predict_data_y_window;
       predictor_info.predict_data_z_window = predictor->_predict_data_z_window;
       predictor_info.x_event_id = predictor->_x_event_id;
       predictor_info.y_event_id = predictor->_y_event_id;
       predictor_info.z_event_id = predictor->_z_event_id;
       predictor_info.queue_check_error = predictor->_queue_check_error;
       predictor_info.max_velocity_x = predictor->_max_velocity_x;
       predictor_info.max_velocity_y = predictor->_max_velocity_y;
       predictor_info.max_velocity_z = predictor->_max_velocity_z;
       predictor_info.running_data_x_t = predictor->_running_data_x_t;
       predictor_info.running_data_y_t = predictor->_running_data_y_t;
       predictor_info.running_data_z_t = predictor->_running_data_z_t;
       predictor_info.running_data_x = predictor->_running_data_x;
       predictor_info.running_data_y = predictor->_running_data_y;
       predictor_info.running_data_z = predictor->_running_data_z;
       predictor->_state_predictor_info_pub.publish(predictor_info);
      }
      //=========================================================end out state predictor information=====================================

    }
    rate.sleep();
  }
  pthread_join(predictor->_predictor_run_thread,NULL);

}

File: DTUAVSIM/Script/Algorithm/Predictor/StatePredictor.cs
using System.Collections;
using System.Collections.Generic;
using System.Threading;
using UnityEngine;
using DTUAVCARS.Algorithm.DataStruct;
namespace DTUAVCARS.Algorithm.Predictor
{
    public struct PredictStateStu
    {
        public List<float> Times;
        public List<float> States;
    }
    public class StatePredictor
    {
        private float _predictorRunHz ;//The Predictor Run Frequency
        private int _saveDataWindow;//The Save Data Window Size
        private int _predictWindow;
        private int _maxPredictWindow;//The Max Data Window Size of Predictor
        private int _minPredictWindow;//The Min Data Window Size of Predictor
        private float _predictTimeDt;//The increated time of Predictor
        private int _predictDataWindow;//The Current Data Window Size of Predictor
        private int _eventId;//The Event Id of Object State
        private int _predictOrder;//The Order of Predicted Function
        private float _queueCheckError;//The Error Value to Check the System State Queue to Use Different Predicted Ways
        private float _maxVelocity;//The Maximun Velocity of Motion
        private CircularBuffer<float> _ciclQueueT;
        private CircularBuffer<float> _ciclQueueX;
        private List<float> _runningDataT;
        private List<float> _runningDataX;
        private LeastSquare _leastSquare;
        private Thread _runningThread;
        private int _runningTime;
        private bool _isStop;
        
        private List<float> _predictStates;
        private List<float> _predictTimes;

        private PredictStateStu _predictStateStu;

       
        public bool SetPredictTimeDt(float timeDt)
        {
            _predictTimeDt = timeDt;
            return true;
        }
        private int GetPredictWindowSize()
        {
            int nowPredictWindowSize = _predictDataWindow;
            if(_eventId == 0)
            {
                nowPredictWindowSize = nowPredictWindowSize + 1;
            }
            else if(_eventId == 1)
            {
                nowPredictWindowSize = nowPredictWindowSize;
            }
            else
            {
                nowPredictWindowSize = _minPredictWindow;
            }

            if(nowPredictWindowSize>_maxPredictWindow)
            {
                nowPredictWindowSize = _maxPredictWindow;
            }
            else if(nowPredictWindowSize<_minPredictWindow)
            {
                nowPredictWindowSize = _minPredictWindow;
            }
            else
            {
                nowPredictWindowSize = nowPredictWindowSize;
            }
            return nowPredictWindowSize;
        }

        private int EventTrigger()
        {
            List<float> devDataQueue = new List<float>();
            int eventId = 0;
            for(int i=1;i<_runningDataX.Count;i++)
            {
                devDataQueue.Add(Mathf.Abs(_runningDataX[i] - _runningDataX[i - 1]));
            }
            float minValue = GetMinValue(_runningDataX);
            float maxValue = GetMaxValue(_runningDataX);
            if(maxValue>_queueCheckError&&minValue>_queueCheckError)
            {
                eventId = 0;
            }
            else if(maxValue<=_queueCheckError&&minValue<=_queueCheckError)
            {
                eventId = 1;
            }
            else
            {
                eventId = 2;
            }

            return eventId;
        }

        private float GetMaxValue(List<float> dataQueue)
        {
            float maxValue = -1000.0f;
            for(int i=0;i<dataQueue.Count;i++)
            {
                if(dataQueue[i]>maxValue)
                {
                    maxValue = dataQueue[i];
                }
            }
            return maxValue;
        }
        private float GetMinValue(List<float> dataQueue)
        {
            float minValue = 1000.0f;
            for(int i=0;i<dataQueue.Count;i++)
            {
                if(dataQueue[i]<minValue)
                {
                    minValue = dataQueue[i];
                }
            }
            return minValue;
        }

        private bool RunLeastSquaresMotion(float currentTime)
        {
            _leastSquare.SetT(_runningDataT);
            _leastSquare.SetX(_runningDataX);
            List<float> result = _leastSquare.GetResult();
            for (int i = 0; i < _predictWindow; i++)
            {
                _predictStateStu.Times[i] = currentTime + (i + 1) * _predictTimeDt;
                for(int j =result.Count-1;j>=0;j--)
                {
                    float predictTime = _runningDataX[_runningDataX.Count - 1] + (i + 1) * _predictTimeDt;
                    _predictStateStu.States[i] = _predictStateStu.States[i] + Mathf.Pow(predictTime, result.Count - 1 - j) * result[j];
                }
            }
            return true;
        }

        private bool RunMaxVelocityMotion(float currentTime)
        {
            float moveDir = _runningDataX[_runningDataX.Count - 1] - _runningDataX[_runningDataX.Count - 2];
            float moveVelocity = 0.0f;
            if(moveDir>0.0f)
            {
                moveVelocity = _maxVelocity;
            }
            else if(moveDir<0.0f)
            {
                moveVelocity = -_maxVelocity;
            }
            else
            {
                moveVelocity = 0.0f;
            }
            float currentState = _runningDataX[_runningDataX.Count - 1];
            for(int i=0;i<_predictWindow;i++)
            {
                _predictStateStu.Times[i] = (i + 1) * _predictTimeDt + currentTime;
                _predictStateStu.States[i] = (i + 1) * moveVelocity * _predictTimeDt + currentState;
            }

            return true;
        }
        private bool RunStaticMotion(float currentTime)
        {
            float currentState = _runningDataX[_runningDataX.Count - 1];
            for(int i=0;i<_predictWindow;i++)
            {
                _predictStateStu.Times[i] = (i + 1) * _predictTimeDt + currentTime;
                _predictStateStu.States[i] = currentState;
            }
            return true;
        }
        private void PredictorRunning()
        {
            while(!_isStop)
            {
                if(_ciclQueueT[_saveDataWindow-1] !=0.0f)
                {
                    float currentTime = _ciclQueueT[_saveDataWindow - 1];
                    float currentPosition = _ciclQueueX[_saveDataWindow - 1];
                    _runningDataT.Clear();
                    _runningDataX.Clear();
                    for(int i=_saveDataWindow-_predictDataWindow;i<_saveDataWindow;i++)
                    {
                        _runningDataT.Add(_ciclQueueT[i] - _ciclQueueT[_saveDataWindow - _predictDataWindow]);
                        _runningDataX.Add(_ciclQueueX[i]);
                    }
                    _eventId = EventTrigger();
                    _predictDataWindow = GetPredictWindowSize();
                    Debug.Log("eventID: " + _eventId);
                    switch(_eventId)
                    {
                        case 0:
                            {
                                RunLeastSquaresMotion(currentTime);
                            }
                            break;
                        case 1:
                            {
                                RunStaticMotion(currentTime);
                            }
                            break;
                        case 2:
                            {
                                RunMaxVelocityMotion(currentTime);
                            }
                            break;
                        default:
                            break;
                    }
                    _predictTimes = _predictStateStu.Times;
                    _predictStates = _predictStateStu.States;
                }

                System.Threading.Thread.Sleep(_runningTime);
            }
        }

        public List<float> GetPredictStates()
        {
            return _predictStates;
        }

        public List<float> GetPredictTimes()
        {
            return _predictTimes;
        }
        public bool StopPredictor()
        {
            _isStop = true;
            return true;
        }
        public bool UpdateDataT(float t)
        {
            _ciclQueueT.Push(t);
            return true;
        }
        public bool UpdateDataX(float x)
        {
            _ciclQueueX.Push(x);
            return true;
        }
        public StatePredictor(float predictorRunHz,int saveDataWindow,int predictWindow, int maxPredictWindow,int minPredictWindow,float predictTimeDt,int predictOrder,float queueCheckError,float maxVelocity,List<float> runningDataT,List<float> runningDataX)
        {
            _predictorRunHz = predictorRunHz;
            _saveDataWindow = saveDataWindow;
            _maxPredictWindow = maxPredictWindow;
            _minPredictWindow = minPredictWindow;
            _predictTimeDt = predictTimeDt;
            _predictOrder = predictOrder;
            _queueCheckError = queueCheckError;
            _maxVelocity = maxVelocity;
            _runningDataT = new List<float>();
            _runningDataT = runningDataT;
            _runningDataX = new List<float>();
            _runningDataX = runningDataX;


            _predictWindow = predictWindow;
            _predictDataWindow = maxPredictWindow;

            _ciclQueueT = new CircularBuffer<float>(_saveDataWindow);
            _ciclQueueX = new CircularBuffer<float>(_saveDataWindow);

            _leastSquare = new LeastSquare(_runningDataT, _runningDataX, _predictOrder);

            _isStop = false;
            _runningTime= (int)((1.0 / _predictorRunHz) * 1000);

            _predictTimes = new List<float>();
            _predictStates = new List<float>();

            _predictStateStu = new PredictStateStu();
            _predictStateStu.States = new List<float>();
            _predictStateStu.Times = new List<float>();

            for (int i = 0; i < _predictWindow; i++)
            {
                _predictTimes.Add(0.0f);
                _predictStates.Add(0.0f);

                _predictStateStu.States.Add(0.0f);
                _predictStateStu.Times.Add(0.0f);
            }

            _runningThread = new Thread(PredictorRunning);
            _runningThread.IsBackground = true;
            _runningThread.Start();

            
        }
    }
}
